Feedbackform: add support for select element placeholder.

samuli opened the pull request:
This adds a new configuration option 'placeholder' for select-elements. When defined, an option without value is added to the beginning of the options-list. If combined with the 'required' attribute, the browser does not accept the placeholder element as a proper selection (due to the empty value-attribute) and prevents the form from being submitted.

As far as I know, the same logic can not be used with optionGroups and therefore 'placeholder' is only implemented for select elements with 'options'.

demiankatz commented:
@samuli, this looks good to me. Just one question: is there a mechanism for adding a placeholder message to text fields (i.e. via the `placeholder` attribute of the `input` or `textarea` tags)? If so, it might make sense to extend support so that you can set placeholders in all the different field types using the same setting; I imagine those types of placeholders could be valuable as well. If it's unreasonably difficult to implement, it's certainly not a requirement, though -- just a thought!

samuli commented:
@demiankatz Textfield and textarea don't support `placeholder`, but one can accomplish the same using `settings`:

```
settings:
   - [placeholder, Please select...]
```
Maybe this should be mentioned in the documentation.

I considered the same mechanism for select-elements, but decided to use a separate attribute to make it clear that the settings-list is appended to the html-tag attributes as such. This is easy to change if needed.

Another option would be to use `placeholder` also for textfield and textarea, but this would only add another way to accomplish the same that can be already done via `settings`.

demiankatz commented:
Thanks, @samuli; I think a note in the comment about the placeholder setting explaining that `settings` can be used for text placeholders should be sufficient!

samuli commented:
@demiankatz Sorry, forgot to check this with radios... will fix.

demiankatz commented:
Thanks, @samuli, I'll wait to hear from you before merging.

samuli commented:
@demiankatz Ok, fixed now. Turns out that only select-element supports option items as keyed arrays (with `label` and `value`). I reverted the handling of options for radios and checkboxes.

demiankatz commented on module/VuFind/src/VuFind/Form/Form.php:
```diff
@@ -245,8 +246,29 @@ protected function parseConfig($formId, $config)
                     foreach ($el['options'] as $option) {
-                        $options[$option] = $this->translate($option);
+                        if ($placeholder) {
+                            $options[] = [
```
I think it's a little confusing that we format $options differently depending on whether or not there is a placeholder. Is this necessary for a particular reason? Would it be better to always format $options in the more verbose way just for consistency/readability? If I'm missing something and we have to do it like this, a comment explaining why would be helpful, so it doesn't get factored out by some future developer. :_)

samuli replied:
Yes, I agree. Refactored to use the more verbose way for select element options also when `placeholder` is not used.

demiankatz replied:
Thanks!
